python/taichi/main.py

```diff
@@ -1,471 +1,174 @@
-    )
-    examples_dir = Path(root_dir) / 'examples'
-    return examples_dir
-
-
-def get_available_examples() -> set:
-    """Get a set of all available example names."""
-    examples_dir = get_examples_dir()
-    all_examples = examples_dir.rglob('*.py')
-    all_example_names = {
-        str(f.resolve()).split('/')[-1].split('.')[0]
-        for f in all_examples
-    }
-    return all_example_names
-
-
-def run_example(name: str):
-    """Run an example based on the example NAME."""
-    all_example_names = get_available_examples()
-    # try to find and remove python file extension once
-    if name.endswith('.py'):
-        name = name.split('.')[0]
-    if name not in all_example_names:
-        sys.exit(
-            f"Sorry, {name} is not an available example name!\nAvailable example names are: {sorted(all_example_names)}"
-        )
-    examples_dir = get_examples_dir()
-    target = str((examples_dir / f"{name}.py").resolve())
-    # we need to modify path for examples that use
-    # implicit relative imports
-    sys.path.append(str(examples_dir.resolve()))
-    print(f"Running example {name} ...")
-    runpy.run_path(target, run_name='__main__')
+    @staticmethod
+    def _get_examples_dir() -> Path:
+        """Get the path to the examples directory."""
+        import taichi as ti
 
+        root_dir = ti.package_root() if ti.is_release(
+        ) else ti.get_repo_directory()
+        examples_dir = Path(root_dir) / 'examples'
+        return examples_dir
+
+    @staticmethod
+    def _get_available_examples() -> set:
+        """Get a set of all available example names."""
+        examples_dir = TaichiMain._get_examples_dir()
+        all_examples = examples_dir.rglob('*.py')
+        all_example_names = {Path(f).stem for f in all_examples}
+        return all_example_names
+
+    @staticmethod
+    def _example_choices_type(choices):
+        def support_choice_with_dot_py(choice):
+            if choice.endswith('.py') and choice.split('.')[0] in choices:
+                # try to find and remove python file extension
+                return choice.split('.')[0]
+            else:
+                return choice
 
-def get_benchmark_baseline_dir():
-    import taichi as ti
-    return os.path.join(ti.core.get_repo_dir(), 'benchmarks', 'baseline')
+        return support_choice_with_dot_py
 
+    @register
+    def example(self, arguments: list = sys.argv[2:]):
+        """Run an example by name (or name.py)"""
+        choices = TaichiMain._get_available_examples()
 
-def get_benchmark_output_dir():
-    import taichi as ti
-    return os.path.join(ti.core.get_repo_dir(), 'benchmarks', 'output')
+        parser = argparse.ArgumentParser(description=f"{self.example.__doc__}")
+        parser.add_argument(
+            "name",
+            help=f"Name of an example (supports .py extension too)\n",
+            type=TaichiMain._example_choices_type(choices),
```

Comment:
This hopefully works similar to what was done in #1098 @TH3CHARLie @archibate

Reply:
Thank you for including this! #1098 would serve as a temporary fix now, this is much more systematic. As @archibate suggested in #1098, using `name[:-3]` would be better to handle corner cases(I believe the corresponding  modification lies in `_example_choices_type`)